File: src/server/agent/stream-event-queue.ts
import type { StreamEventType, StreamModeName, StreamEvent } from "@shared/types/agent";

type EmitFn = (
  type: StreamEventType,
  data: string,
  mode: StreamModeName,
  meta?: StreamEvent["meta"]
) => void;

interface QueuedEvent {
  type: StreamEventType;
  data: string;
  mode: StreamModeName;
  meta: StreamEvent["meta"];
  /** If true, inject insertBefore pointing to the first response part marker */
  wantsInsertBefore: boolean;
}

/**
 * Deferred event queue for the chat stream normalizer.
 *
 * Problem: LangGraph emits tool_call / tool_result from the `updates` channel
 * *after* the response tokens have already streamed via the `messages` channel.
 * This causes tool blocks to appear below the response text in the UI.
 *
 * Solution: defer tool_call / tool_result events from `updates`, then drain
 * them at flush() time with `meta.insertBefore` set to the stable marker ID
 * of the first response part emitted this turn. The frontend uses this to
 * reorder the parts correctly.
 *
 * Marker convention:
 *   The first "response" SSE event of each turn carries
 *   `meta.turnRunId` which is also used as the insertBefore anchor.
 *   Specifically: `insertBefore = "first-response-{turnRunId}"`
 *   and the first response event carries `meta.firstResponseMarker = same value`.
 *   The hook in use-agent-chat.ts maps that marker → the created TextPart id.
 */
export class StreamEventQueue {
  private queue: QueuedEvent[] = [];
  /** Marker string used as insertBefore anchor for this turn */
  private firstResponseMarker: string | null = null;

  /**
   * Called by the normalizer when the first "response" text is about to be emitted.
   * Returns a stable marker string that should be placed in meta.firstResponseMarker.
   */
  setFirstResponseMarker(turnRunId: string): string {
    if (!this.firstResponseMarker) {
      this.firstResponseMarker = `first-response-${turnRunId}`;
    }
    return this.firstResponseMarker;
  }

  /** Whether a first-response marker has been established this turn. */
  hasFirstResponseMarker(): boolean {
    return this.firstResponseMarker !== null;
  }

  /**
   * Enqueue a deferred event. If wantsInsertBefore is true, the event will
   * receive `meta.insertBefore` pointing to the first response marker at drain time.
   */
  enqueue(
    type: StreamEventType,
    data: string,
    mode: StreamModeName,
    meta: StreamEvent["meta"],
    wantsInsertBefore: boolean
  ): void {
    this.queue.push({ type, data, mode, meta, wantsInsertBefore });
  }

  /**
   * Emit all queued events, injecting insertBefore where requested.
   */
  drain(emit: EmitFn): void {
    for (const item of this.queue) {
      const finalMeta: StreamEvent["meta"] =
        item.wantsInsertBefore && this.firstResponseMarker
          ? { ...item.meta, insertBefore: this.firstResponseMarker }
          : item.meta;
      emit(item.type, item.data, item.mode, finalMeta);
    }
    this.queue = [];
  }

  /** Reset for a new turn. */
  reset(): void {
    this.queue = [];
    this.firstResponseMarker = null;
  }
}

File: python/omnimind_backend/runtime/stream_event_queue.py
from dataclasses import dataclass

from omnimind_backend.schemas.agent import StreamEventMeta, StreamEventType, StreamModeName


@dataclass
class QueuedEvent:
    event_type: StreamEventType
    data: str
    mode: StreamModeName
    meta: StreamEventMeta | None
    wants_insert_before: bool


class StreamEventQueue:
    def __init__(self) -> None:
        self._queue: list[QueuedEvent] = []
        self._first_response_marker: str | None = None

    def set_first_response_marker(self, turn_run_id: str) -> str:
        if self._first_response_marker is None:
            self._first_response_marker = f"first-response-{turn_run_id}"
        return self._first_response_marker

    def has_first_response_marker(self) -> bool:
        return self._first_response_marker is not None

    def enqueue(
        self,
        event_type: StreamEventType,
        data: str,
        mode: StreamModeName,
        meta: StreamEventMeta | None,
        wants_insert_before: bool,
    ) -> None:
        self._queue.append(
            QueuedEvent(
                event_type=event_type,
                data=data,
                mode=mode,
                meta=meta,
                wants_insert_before=wants_insert_before,
            )
        )

    def drain(self) -> list[QueuedEvent]:
        drained: list[QueuedEvent] = []
        for item in self._queue:
            if item.wants_insert_before and self._first_response_marker is not None:
                meta = item.meta or StreamEventMeta()
                meta.insertBefore = self._first_response_marker
                item.meta = meta
            drained.append(item)
        self._queue = []
        return drained

    def reset(self) -> None:
        self._queue = []
        self._first_response_marker = None
